lib/signs_ui_web/controllers/auth_controller.ex

```diff
@@ -0,0 +1,25 @@
+defmodule SignsUiWeb.AuthController do
+  use SignsUiWeb, :controller
+  plug(Ueberauth)
+
+  @spec callback(Plug.Conn.t(), map()) :: Plug.Conn.t()
+  def callback(%{assigns: %{ueberauth_auth: auth}} = conn, _params) do
+    username = auth.uid
+    expiration = auth.credentials.expires_at
+
+    current_time = DateTime.utc_now() |> DateTime.to_unix()
+
+    conn
+    |> Guardian.Plug.sign_in(
```

Comment:
Does `uberauth_cognito` take care of verifying that the callback from Cognito is valid, or is that something we'd need to do here?

Reply:
Looks like the answer is `uberauth_cognito` should verify the callback, but currently does not.

Reply:
Yeah, I agree it's `ueberauth_cognito`'s responsibility. What do you mean by verify the callback? There's a few parts to the process. I would say that it's mostly verified. We have a TODO on one part, which we do plan on doing, but I don't believe applies to our use case anyway.

When a request comes into signs-ui's `/auth/cognito/callback` route, it expects a `code=[UUID]` parameter. We then take that code, on the backend, and POST it to Cognito's `/oauth2/token` endpoint. The code must be valid, or else Cognito won't return the access token and JWT profile "bearer token".

The JWT we get back in that request includes a header/signature and could be validated, which we have a TODO on. However, at this point, I believe we _know_ it's valid, since we made the request ourselves to AWS over a TLS connection. As I understand it, the reason for the header/signature bit, is JWT's are meant to be able to be passed around from service to service, and any given service should be able to validate that it's authentic. In our scenario, we got it ourselves over TLS, so we know it's authentic, and immediately use Guardian to create/sign a new token which we put in the session.

One further validation/security mechanism that we have not currently implemented is PKCE. But this one, as I understand it, has a threat model relevant to native apps, where a malicious app running on the device can obtain the initial callback `code` and beat your own app to sending it to the oauth server to get the access token. While it's nice to have, I don't believe it applies to us, with a single app running in our virtualized container.

Reply:
Oh, if you mean the response we get back from Cognito's `/oauth2/token` endpoint, I just realized the version in the repo doesn't have a case clause for anything other than a successful response, so rather than the app getting an authentication failure it just blows up. I have that fixed in my local branch but hadn't pushed that up yet.

Reply:
Can you open a PR for that when it's going in? It's harder to review the code once it's in the master branch.

Reply:
Good point, we'll do an actual PR for that.